### PR TITLE
added orca parser to arkane

### PR DESCRIPTION
**Pull request by dranasinghe**
add OrcaLog and orca test to Arkane. currently limited to parsing energy, geometry, number of atoms, t1 diagnostic.  Freq, spin multiplicity, scan, conformers are not implemented

### CODE REVIEW

File: arkane/orca.py [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+        raise LogError('Could not find a successfully load Orca scan:Parser is not inpemented')
+
+    def load_negative_frequency(self):
+        """
+        Return the imaginary frequency from a transition state frequency
+        calculation in cm^-1.
+        """
+        frequency = 0
+        with open(self.path, 'r') as f:
+            for line in f:
```
[the remark this answers is not in the document]
Comment (mliu49) [reply arkane/orca.py]: `for line in f` is better because it treats f as a iterator, whereas `f.readlines()` will read the entire file first which is less efficient.

File: arkane/orca.py [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+                if 'CARTESIAN COORDINATES (ANGSTROEM)' in line and natoms == 0:
+                    for i in range(2):
+                        line = f.readline()
+
+                    while '---------------------------------' not in line:
+                        natoms += 1
+                        line = f.readline()
+                        if not line.strip():
+                            return natoms
+                line = f.readline()
```
[the remark this answers is not in the document]
Comment (dranasinghe) [reply arkane/orca.py]: Similar to Qchem 0 will be returned if ther is a problem.

**Comment by dranasinghe** [comment]
Thank you for the comments I have addressed all the issues.

**Comment by alongd** [comment]
@dranasinghe, the tests did not pass, giving the following trace:
```
======================================================================
ERROR: test_arkane_examples (arkane.mainTest.TestArkaneExamples)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/ReactionMechanismGenerator/RMG-Py/arkane/mainTest.py", line 70, in test_arkane_examples
    arkane.execute()
  File "/home/travis/build/ReactionMechanismGenerator/RMG-Py/arkane/main.py", line 208, in execute
    job.execute(output_directory=self.output_directory, plot=self.plot, pdep=is_pdep(self.job_list))
  File "/home/travis/build/ReactionMechanismGenerator/RMG-Py/arkane/statmech.py", line 207, in execute
    self.load(pdep, plot)
  File "/home/travis/build/ReactionMechanismGenerator/RMG-Py/arkane/statmech.py", line 331, in load
    energy_log = determine_qm_software(energy.path)
  File "/home/travis/build/ReactionMechanismGenerator/RMG-Py/arkane/util.py", line 67, in determine_qm_software
    line = f.readline()
ValueError: I/O operation on closed file.
```

Let's talk and see how to resolve it.

File: arkane/orca.py [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+                # Automatically determine the number of atoms
+                if 'CARTESIAN COORDINATES (ANGSTROEM)' in line and natoms == 0:
+                    for i in range(2):
+                        line = f.readline()
+
+                    while '---------------------------------' not in line:
+                        natoms += 1
+                        line = f.readline()
+                        if not line.strip():
+                            return natoms
```
Comment (alongd) [inline arkane/orca.py]: I'm not 100% certain, but I think that if we return within an `with open` block, we should close the file before returning

File: arkane/orca.py [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+        Return the number of atoms in the molecular configuration used in
+        the Orca output file.
+        """
+        natoms = 0
+
+        with open(self.path, 'r') as f:
+            line = f.readline()
+            while line != '' and natoms == 0:
+                # Automatically determine the number of atoms
+                if 'CARTESIAN COORDINATES (ANGSTROEM)' in line and natoms == 0:
```
[the remark this answers is not in the document]
Comment (alongd) [reply arkane/orca.py]: Could you explain why this is not needed?

File: arkane/orca.py [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+
+        """
+        Return the optimum geometry of the molecular configuration from the
+        Orca log file. If multiple such geometries are identified, only the
+        last is returned.
+        """
+        atom, coord, number, mass = [], [], [], []
+
+        with open(self.path) as f:
+            log = f.read().splitlines()
```
[the remark this answers is not in the document]
Comment (alongd) [reply arkane/orca.py]: Did you agree with the above comment? I didn't see a respective fix

File: arkane/orca.py [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+                break
+
+        if not completed_job:
+            raise LogError(
+                'Could not find a successfully completed Orca job in Orca output file {0}'.format(self.path))
+
+        # Now look for the geometry.
+        # Will return the final geometry in the file under Standard Nuclear Orientation.
+        geometry_flag = False
+        for i in reversed(range(len(log))):
```
[the remark this answers is not in the document]
Comment (alongd) [reply arkane/orca.py]: Do you agree with the above comment?

File: arkane/orca.py [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+            line = log[i]
+            if 'CARTESIAN COORDINATES (ANGSTROEM)' in line:
+                for line in log[(i + 3):]:
```
[the remark this answers is not in the document]
Comment (alongd) [reply arkane/orca.py]: I think the above comment was not addressed

File: arkane/orca.py [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+        """
+        Load the molecular degree of freedom data from a log file created as
+        the result of a Orca "Freq" quantum chemistry calculation. As
+        CAUTION: The rotational entropy is not quite correctly treated here
+         because it includes a symmetry number that is not yet correctly
+         implemented in ORCA. Orca does not provide Rotational temperatures. Only E(rot) and E(trans) energies
+         Run gaussian to get E0
+        """
+
+        raise LogError('Could not find a successfully load Orca scan:Parser is not inpemented')
```
[the remark this answers is not in the document]
Comment (alongd) [reply arkane/orca.py]: I think the above comment was not addressed

File: arkane/orca.py [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+
+    def load_energy(self, zpe_scale_factor=1.):
+        """
+        Load the energy in J/ml from a Orca  log file. Only the last energy
```
[the remark this answers is not in the document]
Comment (alongd) [reply arkane/orca.py]: I think the above comment was not addressed

File: arkane/orca.py [hunk 10]
```diff
@@ -0,0 +1,229 @@
...
+        Load the unscaled zero-point energy in J/mol from a QChem output file.
+        """
+        zpe = None
+        with open(self.path, 'r') as f:
+            for line in f:
+                if 'Zero point energy' in line:
+                    zpe = float(line.split()[-4]) * constants.E_h * constants.Na
+        if zpe is not None:
+            return zpe
+        else:
```
[the remark this answers is not in the document]
Comment (alongd) [reply arkane/orca.py]: I think the above comment was not addressed

File: arkane/orca.py [hunk 11]
```diff
@@ -0,0 +1,229 @@
...
+            raise LogError('Unable to find zero-point energy in Orca output file.')
+
+    def load_scan_energies(self):
+
+        raise LogError('Could not find a successfully load Orca scan:Parser is not inpemented')
```
[the remark this answers is not in the document]
Comment (alongd) [reply arkane/orca.py]: I think the above comment was not addressed

File: arkane/orca.py [hunk 12]
```diff
@@ -0,0 +1,229 @@
...
+        If multiple occurrences exist, returns the last occurence
+        """
+        with open(self.path) as f:
+            log = f.readlines()
+
+        for line in reversed(log):
+            if 'T1 diagnostic ' in line:
+                items = line.split()
+                return float(items[-1])
+        raise ValueError('Unable to find T1 diagnostic in energy file: {}'.format(self.path))
```
[the remark this answers is not in the document]
Comment (alongd) [reply arkane/orca.py]: I think the above comment was not addressed

File: arkane/orca.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (dranasinghe) [reply arkane/orca.py]: sorry, I missed this the first time. I addressed it now.

File: arkane/orca.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (alongd) [reply arkane/orca.py]: I didn't see any response to this comment. Could you take a look?

File: arkane/orca.py [hunk 13]
```diff
@@ -115,30 +115,30 @@ def load_geometry(self):
         for i in reversed(range(len(log))):
             line = log[i]
             if 'CARTESIAN COORDINATES (ANGSTROEM)' in line:
-                for line in log[(i + 3):]:
-                    if not line.strip():
+                for coordinates in log[(i + 2):]:
```
Comment (alongd) [inline arkane/orca.py]: I think this should remain `line` - otherwise it'll be confusing for someone else to read this code, they would expect coordinates to contain actual coordinates (list or tuple), and not a string with additional info

File: arkane/orca.py [hunk 14]
```diff
@@ -150,11 +150,11 @@ def load_conformer(self, symmetry=None, spin_multiplicity=0, optical_isomers=Non
          Consider using another supported software (such as Gaussian or QChem) to calculate the frequencies and derive E0
         """
 
-        raise LogError('Could not find a successfully load Orca scan:Parser is not inpemented')
+        raise NotImplementedError('Could not find a successfully load Orca scan:Parser is not inpemented')
 
     def load_energy(self, zpe_scale_factor=1.):
         """
-        Load the energy in J/ml from a Orca  log file. Only the last energy
+        Load the energy in J/ml from an Orca  log file. Only the last energy
```
Comment (alongd) [inline arkane/orca.py]: Could you remove the extra space after `Orca`?

File: arkane/orca.py [hunk 15]
```diff
@@ -0,0 +1,231 @@
...
+        Return the number of atoms in the molecular configuration used in
+        the Orca output file.
+        """
+        natoms = 0
+
+        with open(self.path, 'r') as f:
+            line = f.readline()
+            while line != '' and natoms == 0:
+                # Automatically determine the number of atoms
+                if 'CARTESIAN COORDINATES (ANGSTROEM)' in line and natoms == 0:
```
Comment (alongd) [inline arkane/orca.py]: Does this work? There seems to be a typo in Angstrom, but I imagine this is copied from the Orca log file. Could you double-check?

File: arkane/orca.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (alongd) [reply arkane/orca.py]: Can you explain the rational for keeping `natoms == 0` in both the `while` and the `if` statements?

File: arkane/orca.py (hunk 3 above)
Comment (alongd) [reply arkane/orca.py]: I didn't see a response to this comment

File: arkane/orca.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (alongd) [reply arkane/orca.py]: I would change this part to:
```python
if zpe is None:
    raise LogError('Unable to find zero-point energy in Orca output file.')
return zpe
```

File: arkane/orca.py (hunk 15 above)
Comment (dranasinghe) [reply arkane/orca.py]: It was copied from Orca.

File: arkane/orca.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (dranasinghe) [reply arkane/orca.py]: I copied from qchem. I think if the parser could n't find the natom it will return 0. further it will stop once it find natoms.  typically natoms are at the beginning of an outputfile.

**Comment by dranasinghe** [comment]
Thank you for the comments.  I addressed all the comments. I left the comments on the original thread.

**Comment by alongd** [comment]
@dranasinghe, take a look at the Codacy report for this branch, it suggests to add some spaces in `=` assignments in orcaTest.py

File: arkane/orca.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (alongd) [reply arkane/orca.py]: It still shows `f.read().splitlines()`, not `f.readlines()`. Am I missing something?

[the remark this answers is not in the document]
Comment (dranasinghe) [reply arkane/orca.py]: When I remove `f.read().splitlines()` I had problem parsing. so lets leave `splitline()`

File: arkane/orca.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (dranasinghe) [reply arkane/orca.py]: I made the change.

File: arkane/orca.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (alongd) [reply arkane/orca.py]: Do you mean when you replaced it with `lines = f.readlines()`? Strange, we use this format frequently. What problem did you have? Would you like us to look at it together?

File: arkane/statmech.py [hunk 16]
```diff
@@ -59,6 +59,7 @@
 from arkane.molpro import MolproLog
 from arkane.output import prettify
 from arkane.qchem import QChemLog
+from arkane.orca import OrcaLog
```
Comment (alongd) [inline arkane/statmech.py]: We sort imports alphabetically, could you relocate the orca import and put it after the molpro one?

File: arkane/util.py [hunk 17]
```diff
@@ -36,6 +36,7 @@
 from arkane.gaussian import GaussianLog
 from arkane.molpro import MolproLog
 from arkane.qchem import QChemLog
+from arkane.orca import OrcaLog
```
Comment (alongd) [inline arkane/util.py]: Same comment re imports order

File: arkane/orca.py [hunk 18]
```diff
@@ -0,0 +1,231 @@
...
+                line = f.readline()
+
+    def load_force_constant_matrix(self):
+        """not implemented in orca"""
+        # Orca print the hessian to .hess file.  you need to provide .hess instead of .log
+
+        raise LogError('The load_force_constant_matrix method is not implemented for Orca Logs')
+
+    def load_geometry(self):
+
```
Comment (alongd) [inline arkane/orca.py]: remove extra line

File: arkane/orca.py [hunk 19]
```diff
@@ -0,0 +1,231 @@
...
+        """
+        Load the molecular degree of freedom data from a log file created as
+        the result of a Orca "Freq" quantum chemistry calculation. As
+        CAUTION: The rotational entropy is not quite correctly treated here
+         because it includes a symmetry number that is not yet correctly
+         implemented in ORCA. Orca does not provide Rotational temperatures. Only E(rot) and E(trans) energies
+         Consider using another supported software (such as Gaussian or QChem) to calculate the frequencies and derive E0
+        """
+
+        raise NotImplementedError('Could not find a successfully load Orca scan:Parser is not inpemented')
```
Comment (alongd) [inline arkane/orca.py]: Could you rephrase this line and line 187?
There's a typo in `implemented`
let's make the format identical to that we have below, e.g. for `load_scan_pivot_atoms`:
`The load_scan_pivot_atoms method is not implemented for Orca Logs`

**Comment by alongd** [comment]
@dranasinghe, I think we're all set. Squash and rebase?
